File: javascript/lambda/LCFunction.js
'use strict';

class LCFunction {
  constructor(parameter, body) {
    this.parameter = parameter;
    this.body = body;
  }

  replace(name, replacement) {
    if (this.parameter === name) return this;
    return new LCFunction(parameter, body.replace(name, replacement));
  }
}

module.exports = {
  LCFunction
};

File: ruby/lambda_parser/LCFunction.rb
#!/usr/bin/env ruby

class LCFunction < Struct.new(:parameter, :body)
  def replace(name, replacement)
    if self.parameter == name
      self
    else
      LCFunction.new(parameter, body.replace(name, replacement))
    end
  end

  def call(argument)
    body.replace(parameter, argument)
  end

  def callable?
		true
	end

	def reducible?
		false
	end

	def to_s
    "-> #{parameter} { #{body} }"
  end

  def inspect
    to_s
  end
end
